File: frontend/src/lib/api/notifications.ts
// 알림 API 클라이언트

import { apiClient } from './client';

export interface NotificationDto {
  id: string;
  type: string;
  title: string;
  body: string | null;
  is_read: boolean;
  extra: Record<string, unknown> | null;
  created_at: string;
}

export interface NotificationListResponse {
  notifications: NotificationDto[];
  total: number;
  unread: number;
}

export interface UnreadCountResponse {
  unread: number;
}

export interface NotificationPreferencesResponse {
  email: Record<string, boolean>;
  in_app: Record<string, boolean>;
}

export type NotificationPreferencesRequest = NotificationPreferencesResponse;

export const listNotifications = (
  onlyUnread?: boolean,
  limit?: number,
  offset?: number,
): Promise<NotificationListResponse> => {
  const params = new URLSearchParams();
  if (onlyUnread) params.set('only_unread', 'true');
  if (limit) params.set('limit', String(limit));
  if (offset) params.set('offset', String(offset));
  const qs = params.toString();
  return apiClient.get<NotificationListResponse>(`/notifications${qs ? `?${qs}` : ''}`);
};

export const getUnreadCount = (): Promise<UnreadCountResponse> =>
  apiClient.get<UnreadCountResponse>('/notifications/unread-count');

export const markRead = (id: string): Promise<void> =>
  apiClient.put<void>(`/notifications/${id}/read`);

export const markAllRead = (): Promise<{ count: number }> =>
  apiClient.put<{ count: number }>('/notifications/read-all');

export const getPreferences = (): Promise<NotificationPreferencesResponse> =>
  apiClient.get<NotificationPreferencesResponse>('/notifications/preferences');

export const updatePreferences = (
  prefs: NotificationPreferencesRequest,
): Promise<NotificationPreferencesResponse> =>
  apiClient.put<NotificationPreferencesResponse>('/notifications/preferences', prefs);

File: backend/app/api/v1/notifications.py
"""알림 API"""

from fastapi import APIRouter, Depends, Query
from sqlalchemy.orm import Session as DBSession

from app.api.deps import get_current_user
from app.core.database import get_db
from app.schemas.notification import (
    NotificationListResponse,
    NotificationPreferencesRequest,
    NotificationPreferencesResponse,
    UnreadCountResponse,
)
from app.services import notification_service

router = APIRouter(prefix="/notifications", tags=["notifications"])


@router.get("/preferences", response_model=NotificationPreferencesResponse)
def get_preferences(
    current_user: dict = Depends(get_current_user),
    db: DBSession = Depends(get_db),
):
    return notification_service.get_preferences(current_user["id"], db)


@router.put("/preferences", response_model=NotificationPreferencesResponse)
def update_preferences(
    payload: NotificationPreferencesRequest,
    current_user: dict = Depends(get_current_user),
    db: DBSession = Depends(get_db),
):
    return notification_service.update_preferences(
        current_user["id"],
        payload.email.model_dump(),
        payload.in_app.model_dump(),
        db,
    )


@router.get("/unread-count", response_model=UnreadCountResponse)
def get_unread_count(
    current_user: dict = Depends(get_current_user),
    db: DBSession = Depends(get_db),
):
    return {"unread": notification_service.unread_count(current_user["id"], db)}


@router.get("", response_model=NotificationListResponse)
def list_notifications(
    only_unread: bool = Query(False),
    limit: int = Query(50, ge=1, le=200),
    offset: int = Query(0, ge=0),
    current_user: dict = Depends(get_current_user),
    db: DBSession = Depends(get_db),
):
    return notification_service.list_notifications(
        current_user["id"], db, only_unread=only_unread, limit=limit, offset=offset
    )


@router.put("/read-all")
def mark_all_read(
    current_user: dict = Depends(get_current_user),
    db: DBSession = Depends(get_db),
):
    count = notification_service.mark_all_read(current_user["id"], db)
    return {"marked": count}


@router.put("/{notification_id}/read")
def mark_read(
    notification_id: str,
    current_user: dict = Depends(get_current_user),
    db: DBSession = Depends(get_db),
):
    notification_service.mark_read(notification_id, current_user["id"], db)
    return {"ok": True}
